registry/distribution/repository.go

```diff
@@ -0,0 +1,674 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package distribution
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/internal/descriptor"
+	"oras.land/oras-go/v2/internal/httputil"
+	"oras.land/oras-go/v2/registry"
+)
+
+// Repository is a HTTP client to a remote repository.
+type Repository struct {
+	// Client is the underlying HTTP client used to access the remote registry.
+	Client *http.Client
+
+	// Reference references the remote repository.
+	Reference registry.Reference
+
+	// PlainHTTP signals the transport to access the remote repository via HTTP
```

Comment:
Shouldn't the these flags be a requst option and not a part of the client? I would expect you can make multiple requests using the same client with different page sizes concurrently.

Reply:
I thought about that. However, options per API call make the API heavy while those options are not used in the most time.
Note: If an option is used very often, then it should be a direct argument.

Therefore, I change the philosophy behinds it, and adapt the design for `http.Client` and `http.Request`. In other words, the `Registry` and the `Repository` are the options to themselves. All struct fields are exposed with nothing private. Meanwhile, `Registry` and `Repository` accepts concurrent calls by desgin. If you want to do query with different page sizes concurrently, try copying the struct as follows.

```golang
reg, err := distribution.NewRegistry("localhost:5000")
check(err)

reg2 := &distribution.Registry{}
*reg2 = *reg
reg2.RepositoryListPageSize = 2

var res, res2 []string
go reg.Repositories(ctx, func(repos []string) error {
	res = append(res, repos...)
	return nil
})
go reg2.Repositories(ctx, func(repos []string) error {
	res2 = append(res2, repos...)
	return nil
})
```

Reply:
Ok so basically reg are just request options with an internal client. Does it share the http client and connection pools internally since it’s based of the same registry instance?

Reply:
Yes. Actually, the connection pools are shared internally if the underlying `http.Transport` is shared.